src/Lean/Elab/App.lean

```diff
@@ -1689,26 +1703,34 @@ where
         withForallBody body k
       else
         k body
-  go (resultType : Expr) (expectedType : Expr) (previousExceptions : Array Exception) : TermElabM Expr := do
+  go (resultType : Expr) (expectedType : Expr) (previousExceptions : Array Exception) : TermElabM (List (Expr × Syntax × List Syntax)) := do
     let resultType ← instantiateMVars resultType
     let resultTypeFn := resultType.cleanupAnnotations.getAppFn
     try
       tryPostponeIfMVar resultTypeFn
       match resultTypeFn.cleanupAnnotations with
       | .const declName .. =>
-        let idNew := declName ++ id.getId.eraseMacroScopes
-        if (← getEnv).contains idNew then
-          mkConst idNew
-        else if let some (fvar, []) ← resolveLocalName idNew then
-          return fvar
+        let env ← getEnv
+        if isInaccessiblePrivateName env declName then
+          throwError "Invalid dotted identifier notation: The declaration `{.ofConstName declName}` is private"
+        -- Recall that the namespace for private declarations is non-private.
+        let fullName := privateToUserName declName ++ id
+        -- Resolve the name without making use of the current namespace, like in `findMethod?`.
+        let candidates := ResolveName.resolveGlobalName env Name.anonymous (← getOpenDecls) fullName
+          |>.filter (fun (_, fieldList) => fieldList.isEmpty)
+          |>.map Prod.fst
+        if !candidates.isEmpty then
+          candidates.mapM fun resolvedName => return (← mkConst resolvedName, ← getRef, [])
+        else if let some (fvar, []) ← resolveLocalName fullName then
+          return [(fvar, ← getRef, [])]
         else
-          throwUnknownIdentifierAt id <| m!"Unknown identifier `{idNew}`"
+          throwUnknownIdentifierAt (← getRef) (declHint := fullName) <| m!"Unknown constant `{.ofConstName fullName}`"
```

Comment:
@jrr6 Any objection to "constant" instead of 'identifier" here?

Also, would you mind taking a look at the error for `isInaccessiblePrivateName` above? Should it have a `.note`? (I'm also not sure if anyone will ever seriously run into it!)